File: src/main/java/ukalus/level/dungeon/buck/Room.java
package ukalus.level.dungeon.buck;

import ukalus.level.Level;
import ukalus.level.Region;
import ukalus.math.Vector2D;

import java.util.LinkedList;
import java.util.List;
import java.util.Random;

public class Room implements Region<Integer> {

  private Random random;
  private Vector2D location;
  private int height;
  private int width;
  private int id;

  public Room(Random random, int height, int width, int id) {
    this.random = random;
    this.height = height;
    this.width = width;
    this.id = id;
  }

  public void place(Level<Integer> level) {
    place(location, level);
  }

  public void place(Vector2D target, Level<Integer> level) {
    int startX = target.getX();
    int startY = target.getY();

    // interior
    for (int x = startX + 1; x < startX + height - 1; x++) {
      for (int y = startY + 1; y < startY + width - 1; y++) {
        level.set(Vector2D.get(x, y), id);
      }
    }

    // corners
    level.set(Vector2D.get(startX, startY), 0);
    level.set(Vector2D.get(startX + height - 1, startY), 0);
    level.set(Vector2D.get(startX, startY + width - 1), 0);
    level.set(Vector2D.get(startX + height - 1, startY + width - 1), 0);

    List<List<Vector2D>> all = new LinkedList<>();

    // west
    List<Vector2D> open = new LinkedList<>();
    for (int i = startX + 1; i < startX + height - 1; i++) {
      if (startY > 0) {
        if (level.get(Vector2D.get(i, startY - 1)) == 0) {
          if (!open.isEmpty()) {
            all.add(open);
            open = new LinkedList<>();
          }
        } else {
          open.add(Vector2D.get(i, startY));
        }
      }
      level.set(Vector2D.get(i, startY), 0);
    }
    if (!open.isEmpty()) {
      all.add(open);
      open = new LinkedList<>();
    }

    // east
    for (int i = startX + 1; i < startX + height - 1; i++) {
      if (startY + width < level.getWidth()) {
        if (level.get(Vector2D.get(i, startY + width)) == 0) {
          if (!open.isEmpty()) {
            all.add(open);
            open = new LinkedList<>();
          }
        } else {
          open.add(Vector2D.get(i, startY + width - 1));
        }
      }
      level.set(Vector2D.get(i, startY + width - 1), 0);
    }
    if (!open.isEmpty()) {
      all.add(open);
      open = new LinkedList<>();
    }

    // north
    for (int i = startY + 1; i < startY + width - 1; i++) {
      if (startX > 0) {
        if (level.get(Vector2D.get(startX - 1, i)) == 0) {
          if (!open.isEmpty()) {
            all.add(open);
            open = new LinkedList<>();
          }
        } else {
          open.add(Vector2D.get(startX, i));
        }
      }
      level.set(Vector2D.get(startX, i), 0);
    }
    if (!open.isEmpty()) {
      all.add(open);
      open = new LinkedList<>();
    }

    // south
    for (int i = startY + 1; i < startY + width - 1; i++) {
      if (startX + height < level.getLength()) {
        if (level.get(Vector2D.get(startX + height, i)) == 0) {
          if (!open.isEmpty()) {
            all.add(open);
            open = new LinkedList<>();
          }
        } else {
          open.add(Vector2D.get(startX + height - 1, i));
        }
      }
      level.set(Vector2D.get(startX + height - 1, i), 0);
    }
    if (!open.isEmpty()) {
      all.add(open);
    }

    // pick door ways
    for (List<Vector2D> list : all) {
      if (!list.isEmpty()) {
        Vector2D location = list.get(random.nextInt(list.size()));
        level.set(Vector2D.get(location.getX(), location.getY()), 100);
      }
    }
  }

  public int cost(Level<Integer> level) {
    return cost(location, level);
  }

  public int cost(Vector2D target, Level<Integer> level) {
    int result = 0;
    int startX = target.getX();
    int startY = target.getY();
    if (startY >= 0
        && startY + width <= level.getWidth()
        && startX >= 0
        && startX + height <= level.getLength()
        && level.get(Vector2D.get(startX, startY)) == 0
        && level.get(Vector2D.get(startX, startY + width - 1)) == 0
        && level.get(Vector2D.get(startX + height - 1, startY)) == 0
        && level.get(Vector2D.get(startX + height - 1, startY + width - 1)) == 0) {

      if (startY > 0 && startY + width < level.getWidth()) {
        for (int x = startX + 1; x < startX + height - 1; x++) {
          if (level.get(Vector2D.get(x, startY - 1)) == 1) {
            result++;
          }
          if (level.get(Vector2D.get(x, startY + width)) == 1) {
            result++;
          }
        }
      }
      if (startX > 0 && startX + height < level.getLength()) {
        for (int y = startY + 1; y < startY + width - 1; y++) {
          if (level.get(Vector2D.get(startX - 1, y)) == 1) {
            result++;
          }
          if (level.get(Vector2D.get(startX + height, y)) == 1) {
            result++;
          }
        }
      }
      for (int x = startX + 1; x < startX + height - 1; x++) {
        for (int y = startY + 1; y < startY + width - 1; y++) {
          if (level.get(Vector2D.get(x, y)) == 1) {
            result += 3;
          }
          if (level.get(Vector2D.get(x, y)) > 1) {
            result += 100;
          }
        }
      }
    }
    return result;
  }

  public Vector2D getLocation() {
    return location;
  }

  public void setLocation(Vector2D location) {
    this.location = location;
  }

}

File: src/main/kotlin/ukalus/level/dungeon/buck/Room.kt
package ukalus.level.dungeon.buck

import ukalus.level.Level
import ukalus.level.Region
import ukalus.math.Vector2D
import ukalus.util.sift
import java.util.*

class Room(
        private val random: Random,
        private val height: Int,
        private val width: Int,
        private val id: Int
) : Region<Int> {
    override fun place(location: Vector2D, level: Level<Int>) {
        val startX = location.x
        val startY = location.y

        // interior
        for (x in startX + 1..startX + height - 1 - 1) {
            for (y in startY + 1..startY + width - 1 - 1) {
                level[Vector2D(x, y)] = id
            }
        }

        // corners
        level[Vector2D(startX, startY)] = 0
        level[Vector2D(startX + height - 1, startY)] = 0
        level[Vector2D(startX, startY + width - 1)] = 0
        level[Vector2D(startX + height - 1, startY + width - 1)] = 0

        // edges
        val borders = listOf(
                (startX + 1 until startX + height - 1).map { Pair(Vector2D(it, startY), Vector2D(it, startY - 1)) },
                (startX + 1 until startX + height - 1).map { Pair(Vector2D(it, startY + width - 1), Vector2D(it, startY + width)) },
                (startY + 1 until startY + width - 1).map { Pair(Vector2D(startX, it), Vector2D(startX - 1, it)) },
                (startY + 1 until startY + width - 1).map { Pair(Vector2D(startX + height - 1, it), Vector2D(startX + height, it)) }
        )
        borders.forEach { items -> items.forEach { level[it.first] = 0 } }

        // select door ways
        borders
                .flatMap { items -> items.sift { level.contains(it.second) && level[it.second] == 0 }.asIterable() }
                .map { it[random.nextInt(it.size)] }
                .forEach { level[it.first] = 100 }
    }

    override fun cost(location: Vector2D, level: Level<Int>): Int {
        var result = 0
        val startX = location.x
        val startY = location.y
        if (startY >= 0
                && startY + width <= level.width
                && startX >= 0
                && startX + height <= level.length
                && level[Vector2D(startX, startY)] == 0
                && level[Vector2D(startX, startY + width - 1)] == 0
                && level[Vector2D(startX + height - 1, startY)] == 0
                && level[Vector2D(startX + height - 1, startY + width - 1)] == 0) {

            if (startY > 0 && startY + width < level.width) {
                for (x in startX + 1..startX + height - 1 - 1) {
                    if (level[Vector2D(x, startY - 1)] == 1) {
                        result++
                    }
                    if (level[Vector2D(x, startY + width)] == 1) {
                        result++
                    }
                }
            }
            if (startX > 0 && startX + height < level.length) {
                for (y in startY + 1..startY + width - 1 - 1) {
                    if (level[Vector2D(startX - 1, y)] == 1) {
                        result++
                    }
                    if (level[Vector2D(startX + height, y)] == 1) {
                        result++
                    }
                }
            }
            for (x in startX + 1..startX + height - 1 - 1) {
                for (y in startY + 1..startY + width - 1 - 1) {
                    if (level[Vector2D(x, y)] == 1) {
                        result += 3
                    }
                    if (level[Vector2D(x, y)] > 1) {
                        result += 100
                    }
                }
            }
        }
        return result
    }
}

